sherpa-ncnn/csrc/context-graph.cc

```diff
@@ -0,0 +1,103 @@
+// sherpa-onnx/csrc/context-graph.cc
+//
+// Copyright (c)  2023  Xiaomi Corporation
+
+#include "sherpa-ncnn/csrc/context-graph.h"
+
+#include <cassert>
+#include <queue>
+#include <utility>
+
+namespace sherpa_ncnn {
+void ContextGraph::Build(
+    const std::vector<std::vector<int32_t>> &token_ids) const {
+  for (int32_t i = 0; i < token_ids.size(); ++i) {
+    auto node = root_.get();
+    for (int32_t j = 0; j < token_ids[i].size(); ++j) {
+      int32_t token = token_ids[i][j];
+      if (0 == node->next.count(token)) {
+        bool is_end = j == token_ids[i].size() - 1;
+        node->next[token] = std::make_unique<ContextState>(
+            token, context_score_, node->node_score + context_score_,
+            is_end ? node->node_score + context_score_ : 0, is_end);
+      }
+      node = node->next[token].get();
+    }
+  }
+  FillFailOutput();
+}
+
+std::pair<float, const ContextState *> ContextGraph::ForwardOneStep(
+    const ContextState *state, int32_t token) const {
+  const ContextState *node;
+  float score;
+  if (1 == state->next.count(token)) {
+    node = state->next.at(token).get();
+    score = node->token_score;
+  } else {
+    node = state->fail;
+    while (0 == node->next.count(token)) {
+      node = node->fail;
+      if (-1 == node->token) break;  // root
+    }
+    if (1 == node->next.count(token)) {
+      node = node->next.at(token).get();
+    }
+    score = node->node_score - state->node_score;
+  }
+//  SHERPA_ONNX_CHECK(nullptr != node);
```

Comment:
Have you finished this PR?  please delete these lines if you don't need them any more.

Reply:
done

Reply:
Great! Thanks! It would be nice if you can post some decoding logs showing the results with/without contexts, just like this https://github.com/k2-fsa/sherpa/pull/407#issuecomment-1597014596 .

Reply:
@pkufool 
without contexts(The content of the hotwords file is empty):
`decode method:modified_beam_search
RecognizerConfig(feat_config=FeatureExtractorConfig(sampling_rate=16000, feature_dim=80), model_config=ModelConfig(encoder_param="sherpa-ncnn-streaming-zipformer-zh-14M-2023-02-23/encoder_jit_trace-pnnx.ncnn.param", encoder_bin="sherpa-ncnn-streaming-zipformer-zh-14M-2023-02-23/encoder_jit_trace-pnnx.ncnn.bin", decoder_param="sherpa-ncnn-streaming-zipformer-zh-14M-2023-02-23/decoder_jit_trace-pnnx.ncnn.param", decoder_bin="sherpa-ncnn-streaming-zipformer-zh-14M-2023-02-23/decoder_jit_trace-pnnx.ncnn.bin", joiner_param="sherpa-ncnn-streaming-zipformer-zh-14M-2023-02-23/joiner_jit_trace-pnnx.ncnn.param", joiner_bin="sherpa-ncnn-streaming-zipformer-zh-14M-2023-02-23/joiner_jit_trace-pnnx.ncnn.bin", tokens="sherpa-ncnn-streaming-zipformer-zh-14M-2023-02-23/tokens.txt", encoder num_threads=1, decoder num_threads=1, joiner num_threads=1), decoder_config=DecoderConfig(method="modified_beam_search", num_active_paths=4), max_active_paths=4, endpoint_config=EndpointConfig(rule1=EndpointRule(must_contain_nonsilence=False, min_trailing_silence=2.4, min_utterance_length=0), rule2=EndpointRule(must_contain_nonsilence=True, min_trailing_silence=1.4, min_utterance_length=0), rule3=EndpointRule(must_contain_nonsilence=False, min_trailing_silence=0, min_utterance_length=20)), enable_endpoint=False, context_score=1.5, decoding_method="")
wav filename: hotwords.wav
wav duration (s): 4.334
Started!
create contexts stream
Done!
Recognition result for hotwords.wav
text: 五百去服罗北以县保护平
timestamps: 0.96 1.12 1.24 1.76 2.24 2.56 2.88 3.04 3.28 3.44 3.88 
Elapsed seconds: 0.488 s
Real time factor (RTF): 0.488 / 4.334 = 0.113`

with contexts, the hotwords file has two string segmented by space(if the language is not something like CJK,  for exampel: English, you need bpe model to segment hotwords manually):
罗 北 以 线
保 护 屏
`decode method:modified_beam_search
RecognizerConfig(feat_config=FeatureExtractorConfig(sampling_rate=16000, feature_dim=80), model_config=ModelConfig(encoder_param="sherpa-ncnn-streaming-zipformer-zh-14M-2023-02-23/encoder_jit_trace-pnnx.ncnn.param", encoder_bin="sherpa-ncnn-streaming-zipformer-zh-14M-2023-02-23/encoder_jit_trace-pnnx.ncnn.bin", decoder_param="sherpa-ncnn-streaming-zipformer-zh-14M-2023-02-23/decoder_jit_trace-pnnx.ncnn.param", decoder_bin="sherpa-ncnn-streaming-zipformer-zh-14M-2023-02-23/decoder_jit_trace-pnnx.ncnn.bin", joiner_param="sherpa-ncnn-streaming-zipformer-zh-14M-2023-02-23/joiner_jit_trace-pnnx.ncnn.param", joiner_bin="sherpa-ncnn-streaming-zipformer-zh-14M-2023-02-23/joiner_jit_trace-pnnx.ncnn.bin", tokens="sherpa-ncnn-streaming-zipformer-zh-14M-2023-02-23/tokens.txt", encoder num_threads=1, decoder num_threads=1, joiner num_threads=1), decoder_config=DecoderConfig(method="modified_beam_search", num_active_paths=4), max_active_paths=4, endpoint_config=EndpointConfig(rule1=EndpointRule(must_contain_nonsilence=False, min_trailing_silence=2.4, min_utterance_length=0), rule2=EndpointRule(must_contain_nonsilence=True, min_trailing_silence=1.4, min_utterance_length=0), rule3=EndpointRule(must_contain_nonsilence=False, min_trailing_silence=0, min_utterance_length=20)), enable_endpoint=False, context_score=1.5, decoding_method="")
wav filename: hotwords.wav
wav duration (s): 4.334
Started!
罗 北 以 线 保 护 屏 create contexts stream
Done!
Recognition result for hotwords.wav
text: 五百去服罗北以线保护屏
timestamps: 0.96 1.12 1.24 1.76 2.24 2.44 2.68 2.92 3.32 3.44 3.72 
Elapsed seconds: 0.447 s
Real time factor (RTF): 0.447 / 4.334 = 0.103`